File: fetch_wiki_images.mjs
import fs from 'fs';

async function getWikiImage(query) {
  const wikiUrl = 'https://en.wikipedia.org/w/api.php?action=query&generator=search&gsrsearch=' + encodeURIComponent(query) + '&gsrlimit=1&prop=pageimages&pithumbsize=600&format=json';
  try {
    const wikiRes = await fetch(wikiUrl);
    const wikiJson = await wikiRes.json();
    if (wikiJson.query && wikiJson.query.pages) {
      const pages = Object.values(wikiJson.query.pages);
      if (pages.length > 0 && pages[0].thumbnail) {
        return pages[0].thumbnail.source;
      }
    }
  } catch (e) {}
  return null;
}

(async () => {
  const content = fs.readFileSync('src/data/products.ts', 'utf8');
  const match = content.match(/const ROWS: Row\[\] = \[([\s\S]*?)\];/);
  if (!match) return;
  const rowsRaw = match[1].split('],\n');
  
  const mappings = {};
  
  for (const row of rowsRaw) {
    const titleMatch = row.match(/"([^"]+)"/);
    if (titleMatch) {
      const title = titleMatch[1];
      
      // Simplify query for better wiki hits
      let query = title.replace(/\(.*?\)/g, '').trim(); // remove (5 kg) etc
      query = query.split(' ').slice(0, 2).join(' '); // take first 2-3 words
      
      let url = await getWikiImage(query);
      if (!url) {
         // fallback to category-like search
         if (title.toLowerCase().includes('seed')) url = await getWikiImage('seeds agriculture');
         else if (title.toLowerCase().includes('fertilizer') || title.toLowerCase().includes('urea')) url = await getWikiImage('fertilizer bag agriculture');
         else if (title.toLowerCase().includes('tractor') || title.toLowerCase().includes('sprayer')) url = await getWikiImage('agricultural machinery');
         else url = await getWikiImage('agriculture');
      }
      
      if (url) {
        mappings[title] = url;
        console.log('Found:', title, '->', url);
      } else {
        mappings[title] = 'https://upload.wikimedia.org/wikipedia/commons/thumb/1/13/Tractor_in_a_field.jpg/600px-Tractor_in_a_field.jpg';
      }
    }
  }
  fs.writeFileSync('src/data/image_mappings.json', JSON.stringify(mappings, null, 2));
  console.log('Done!');
})();

File: fetch_wiki_images.py
import json
import urllib.request
import urllib.parse
import time

with open('product_list.json', 'r', encoding='utf-8') as f:
    products = json.load(f)

# Load existing successful mappings
try:
    with open('image_mappings.json', 'r', encoding='utf-8') as f:
        mappings = json.load(f)
except:
    mappings = {}

def get_wiki_image(search_term):
    # Clean the search term
    # E.g., "Paddy Seed PR-126 (5 kg)" -> "Paddy"
    clean = search_term.split('(')[0].strip()
    words = clean.split()
    # take first two words usually to get the essence (e.g., "Paddy Seed", "Urea Fertilizer")
    if len(words) > 2:
        clean = " ".join(words[:2])
        
    url = f"https://en.wikipedia.org/w/api.php?action=query&list=search&srsearch={urllib.parse.quote(clean)}&srnamespace=6&format=json&utf8="
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'PureFarm-Bot/1.0'})
        res = urllib.request.urlopen(req).read()
        data = json.loads(res)
        for s in data['query']['search'][:1]:
            title = s['title']
            iurl = f"https://en.wikipedia.org/w/api.php?action=query&titles={urllib.parse.quote(title)}&prop=imageinfo&iiprop=url&format=json"
            ireq = urllib.request.Request(iurl, headers={'User-Agent': 'PureFarm-Bot/1.0'})
            ires = urllib.request.urlopen(ireq).read()
            idata = json.loads(ires)
            pages = idata['query']['pages']
            for p in pages:
                if 'imageinfo' in pages[p]:
                    img_url = pages[p]['imageinfo'][0]['url']
                    if not img_url.endswith('.svg') and not img_url.endswith('.pdf'):
                        return img_url
    except Exception as e:
        print("Error fetching from Wiki:", e)
    return None

for i, p in enumerate(products):
    name = p['name']
    if name not in mappings:
        print(f"[{i+1}/120] Fetching wiki for {name}...")
        img = get_wiki_image(name)
        if img:
            mappings[name] = img
        time.sleep(0.5)

with open('image_mappings.json', 'w', encoding='utf-8') as f:
    json.dump(mappings, f, indent=2)

print(f"Total mappings now: {len(mappings)}/120")
